### PR TITLE
add MAEP node

### PR DESCRIPTION
**Pull request by UrbanGISer**
add MAEP node in location analysis

### CODE REVIEW

**Review by koettert — CHANGES_REQUESTED** [review]

File: knime_extension/src/nodes/locationanalysis.py [hunk 1]
```diff
@@ -635,3 +635,202 @@ def getIndex(s):
...
+@knext.node(
+    name="MAEP Solver",
+    node_type=knext.NodeType.PREDICTOR,
+    icon_path=__NODE_ICON_PATH + "MAEP.png",
+    category=__category,
+    after="",
+)
+@knext.input_table(
+    name="Input tabale for demand and new facilities",
+    description="""The table MUST contain m rows and (k+1) columns,
```
[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/locationanalysis.py:651]: added

File: knime_extension/src/nodes/locationanalysis.py [hunk 2]
```diff
@@ -635,3 +635,202 @@ def getIndex(s):
...
+    name="Input capacity table for existing facilities ",
+    description="""The table provides information on the supply capacity of each existing facility.
+    This table can be used in conjunction with the other input tables to determine the optimal location of 
+    new facilities while taking into account the capacity constraints of existing facilities.
+    Please ensure that the order of rows in the capacity column is consistent with the order of columns OD Matrix for Existing Facilities.
+    """,
+)
+@knext.output_table(
+    name="P-median result table",
+    description="candidate column name and chosen status(1/0)",
```
[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/locationanalysis.py:673]: revised

File: knime_extension/src/nodes/locationanalysis.py [hunk 3]
```diff
@@ -635,3 +635,202 @@ def getIndex(s):
...
+)
+@knut.pulp_node_description(
+    short_description="maximal accessibility equality problem (MAEP)",
```
[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/locationanalysis.py:676]: revised

File: knime_extension/src/nodes/locationanalysis.py [hunk 4]
```diff
@@ -635,3 +635,202 @@ def getIndex(s):
...
+    description="""The optimization objective of MAEP is to minimize inequality in accessibility of facilities
```
[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/locationanalysis.py:677]: added

File: knime_extension/src/nodes/locationanalysis.py [hunk 5]
```diff
@@ -635,3 +635,202 @@ def getIndex(s):
...
+    it becomes a nonlinear programming (NLP) or quadratic programming (QP).
+    The MAEP problem will be solved by cvxopt package. 
+    """,
+    references={
+        "cvxopt": "https://cvxopt.org/",
+    },
+)
+class MAEPSolverNode:
+
+    Newcapacity = knext.DoubleParameter(
```
[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/locationanalysis.py:688]: revised

File: knime_extension/src/nodes/locationanalysis.py [hunk 6]
```diff
@@ -635,3 +635,202 @@ def getIndex(s):
...
+        "Input new capacity",
+        "New resource or capacility for facilities .",
```
[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/locationanalysis.py]: revised

File: knime_extension/src/nodes/locationanalysis.py [hunk 7]
```diff
@@ -635,3 +635,202 @@ def getIndex(s):
...
+    )
+    Demand = knext.ColumnParameter(
+        "Demand column",
+        "The column for demand(e.g.,population). ",
```
[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/locationanalysis.py]: revised

File: knime_extension/src/nodes/locationanalysis.py [hunk 8]
```diff
@@ -635,3 +635,202 @@ def getIndex(s):
...
+        port_index=0,
+        column_filter=knut.is_numeric,
+        include_row_key=False,
+        include_none_column=False,
+    )
+
+    Supply = knext.ColumnParameter(
+        "Column for supply facility",
+        "Travel cost column of  required facility or Travel cost column of nearest required facilities. ",
```
[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/locationanalysis.py]: revised

File: knime_extension/src/nodes/locationanalysis.py [hunk 9]
```diff
@@ -635,3 +635,202 @@ def getIndex(s):
...
+        "Distance decay parameters",
+        "It works for the parameters for the chosen corresponding models",
+    )
+
+    def configure(
+        self, configure_context, input_schema_1, input_schema_2, input_schema_3
+    ):
+        return knext.Schema.from_columns(
+            [
+                knext.Column(knext.string(), "FacilityID"),
```
[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/locationanalysis.py]: revised

File: knime_extension/src/nodes/locationanalysis.py [hunk 10]
```diff
@@ -635,3 +635,202 @@ def getIndex(s):
...
+    def execute(self, exec_context: knext.ExecutionContext, input_1, input_2, input_3):
+        df1 = gp.GeoDataFrame(input_1.to_pandas()).reset_index(drop=True)
+        df2 = gp.GeoDataFrame(input_2.to_pandas()).reset_index(drop=True)
+        df3 = gp.GeoDataFrame(input_3.to_pandas()).reset_index(drop=True)
+        import numpy as np
+        import math
+        import cvxopt
+        import pandas as pd
+
+        D = df1[self.Demand]
```
[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/locationanalysis.py]: revised

File: knime_extension/src/nodes/locationanalysis.py [hunk 11]
```diff
@@ -635,3 +635,202 @@ def getIndex(s):
...
+    )
+    Decaypara = knext.DoubleParameter(
+        "Distance decay parameters",
+        "It works for the parameters for the chosen corresponding models",
+    )
+
+    def configure(
+        self, configure_context, input_schema_1, input_schema_2, input_schema_3
+    ):
+        return knext.Schema.from_columns(
```
[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/locationanalysis.py:763]: they can not be preset

**Review by koettert — CHANGES_REQUESTED** [review]
Please upload a test workflow for the node to the KNIME Hub otherwise I cannot test it.

File: knime_extension/src/nodes/locationanalysis.py [hunk 12]
```diff
@@ -635,3 +635,202 @@ def getIndex(s):
...
+    Supply = knext.ColumnParameter(
+        "Column for supply facility",
+        "Travel cost column of  required facility or Travel cost column of nearest required facilities. ",
+        port_index=2,
+        column_filter=knut.is_numeric,
+        include_row_key=False,
+        include_none_column=False,
+    )
+
+    Decaymodel = knext.StringParameter(
```
[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/locationanalysis.py]: revised

File: knime_extension/src/nodes/locationanalysis.py [hunk 13]
```diff
@@ -635,3 +635,202 @@ def getIndex(s):
...
+        return knext.Schema.from_columns(
+            [
+                knext.Column(knext.string(), "FacilityID"),
+                knext.Column(knext.double(), "All"),
+                knext.Column(knext.double(), "Fixed"),
+            ]
+        )
+
+    def execute(self, exec_context: knext.ExecutionContext, input_1, input_2, input_3):
+        df1 = gp.GeoDataFrame(input_1.to_pandas()).reset_index(drop=True)
```
[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/locationanalysis.py]: revised

File: knime_extension/src/nodes/locationanalysis.py [hunk 14]
```diff
@@ -635,3 +635,202 @@ def getIndex(s):
...
+############################################
+@knext.node(
+    name="MAEP Solver",
+    node_type=knext.NodeType.PREDICTOR,
+    icon_path=__NODE_ICON_PATH + "MAEP.png",
+    category=__category,
+    after="",
+)
+@knext.input_table(
+    name="Input tabale for demand and new facilities",
```
[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/locationanalysis.py]: revised

File: knime_extension/src/nodes/locationanalysis.py [hunk 15]
```diff
@@ -633,3 +628,260 @@ def getIndex(s):
...
+        return knext.Schema.from_columns(
+            [
+                knext.Column(knext.string(), "Facility ID"),
+                knext.Column(knext.double(), "All"),
+                knext.Column(knext.double(), "Fixed"),
+            ]
+        )
+
+    def execute(self, exec_context: knext.ExecutionContext, input_1, input_2, input_3):
+        dfn1 = gp.GeoDataFrame(input_1.to_pandas()).reset_index(drop=True)
```
[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/locationanalysis.py:779]: same as the previous location analysis node,  the data can use geo or non-geo data

[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/locationanalysis.py:779]: Is there any other way to import the data

**Comment by koettert** [comment]
Changes were integrated into the other branch and merged to main

**Pull request closed without merging**